File: src/incremental_processor.py
"""
Incremental PHPStan Error Processor

This module provides functionality to incrementally process PHPStan errors
with Claude Code via MCP, processing a few errors at a time until all are fixed.
"""

import os
import sys
import time
import json
from typing import Dict, List, Optional, Any, Tuple

# Add the project root to the Python path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from src.PhpStanRunner.Runner import PhpStanRunner
from src.ErrorFormatter.Formatter import PhpStanErrorFormatter
from src.McpIntegration.McpClient import McpClient


class IncrementalProcessor:
    """Processes PHPStan errors incrementally with Claude Code via MCP."""

    def __init__(self, 
                 project_path: str,
                 max_errors_per_batch: int = 3,
                 max_iterations: int = 10,
                 phpstan_binary: Optional[str] = None,
                 verbose: bool = False):
        """
        Initialize the incremental processor.
        
        Args:
            project_path: Path to the PHP project
            max_errors_per_batch: Maximum number of errors to process in a single batch
            max_iterations: Maximum number of iterations to run
            phpstan_binary: Path to the PHPStan binary
            verbose: Whether to run in verbose mode
        """
        self.project_path = os.path.abspath(project_path)
        self.max_errors_per_batch = max_errors_per_batch
        self.max_iterations = max_iterations
        self.verbose = verbose
        
        # Initialize MCP client
        self.mcp_client = McpClient(project_path, max_errors_per_batch, phpstan_binary)
        
        # Initialize PHPStan runner
        self.phpstan_runner = PhpStanRunner(project_path, phpstan_binary)
        
        # Initialize error formatter
        self.error_formatter = PhpStanErrorFormatter(max_errors_per_batch)
        
        # Statistics
        self.stats = {
            "iterations": 0,
            "total_errors_fixed": 0,
            "total_batches_processed": 0,
            "errors_by_type": {}
        }

    def process(self, 
               paths: Optional[List[str]] = None,
               level: Optional[int] = None,
               config_path: Optional[str] = None) -> bool:
        """
        Process PHPStan errors incrementally.
        
        Args:
            paths: List of paths to analyze
            level: PHPStan rule level (0-9)
            config_path: Path to PHPStan config file
            
        Returns:
            True if all errors were fixed, False otherwise
        """
        iteration = 0
        total_errors_fixed = 0
        
        while iteration < self.max_iterations:
            print(f"\n=== Iteration {iteration + 1}/{self.max_iterations} ===")
            
            # Run PHPStan analysis
            output, return_code, batches = self.mcp_client.run_phpstan_analysis(
                paths, level, config_path, self.verbose
            )
            
            # If no errors, we're done
            if return_code == 0:
                print(f"\n✅ All PHPStan errors fixed! Total fixed: {total_errors_fixed}")
                self.stats["iterations"] = iteration + 1
                self.stats["total_errors_fixed"] = total_errors_fixed
                return True
            
            # Get total errors in this iteration
            total_errors = sum(batch["batch"]["batch_size"] for batch in batches)
            print(f"Found {total_errors} errors in {len(batches)} batches")
            
            # Process batches one by one
            errors_fixed_this_iteration = 0
            
            for batch_idx, batch in enumerate(batches):
                print(f"\n--- Processing batch {batch_idx + 1}/{len(batches)} ---")
                batch_size = batch["batch"]["batch_size"]
                
                # Prepare MCP message with file contents
                file_contents = self._get_file_contents(batch["errors_by_file"].keys())
                message = self.mcp_client.prepare_mcp_message(batch, file_contents)
                
                # Send to Claude
                print(f"Sending batch with {batch_size} errors to Claude...")
                response = self.mcp_client.send_to_claude(message)
                
                # Apply fixes
                if "fixes" in response and response["fixes"]:
                    num_fixes = len(response["fixes"])
                    print(f"Applying {num_fixes} fixes suggested by Claude...")
                    
                    success = self.mcp_client.apply_fixes(response["fixes"])
                    if success:
                        errors_fixed_this_iteration += num_fixes
                        self._update_error_stats(batch)
                    else:
                        print("❌ Failed to apply fixes")
                        return False
                else:
                    print("⚠️ No fixes suggested by Claude for this batch")
                
                # Update statistics
                self.stats["total_batches_processed"] += 1
                
                # Optional: Add a delay between batches to avoid rate limiting
                if batch_idx < len(batches) - 1:
                    time.sleep(1)
            
            # Update total errors fixed
            total_errors_fixed += errors_fixed_this_iteration
            
            # If no errors were fixed in this iteration, we're stuck
            if errors_fixed_this_iteration == 0:
                print(f"\n⚠️ No errors fixed in iteration {iteration + 1}. Stopping.")
                self.stats["iterations"] = iteration + 1
                self.stats["total_errors_fixed"] = total_errors_fixed
                return False
            
            print(f"\nFixed {errors_fixed_this_iteration} errors in iteration {iteration + 1}")
            print(f"Total errors fixed so far: {total_errors_fixed}")
            
            # Increment iteration counter
            iteration += 1
        
        # If we've reached max_iterations, we've failed to fix all errors
        print(f"\n⚠️ Reached maximum iterations ({self.max_iterations}) without fixing all errors")
        print(f"Total errors fixed: {total_errors_fixed}")
        
        self.stats["iterations"] = self.max_iterations
        self.stats["total_errors_fixed"] = total_errors_fixed
        
        return False

    def _get_file_contents(self, file_paths: List[str]) -> Dict[str, str]:
        """
        Get the contents of the specified files.
        
        Args:
            file_paths: List of file paths
            
        Returns:
            Dictionary mapping file paths to their contents
        """
        file_contents = {}
        
        for file_path in file_paths:
            full_path = os.path.join(self.project_path, file_path)
            if os.path.isfile(full_path):
                try:
                    with open(full_path, 'r') as f:
                        file_contents[file_path] = f.read()
                except Exception as e:
                    print(f"Error reading file {file_path}: {str(e)}")
        
        return file_contents

    def _update_error_stats(self, batch: Dict[str, Any]) -> None:
        """
        Update error statistics.
        
        Args:
            batch: Batch of errors
        """
        for file_path, errors in batch["errors_by_file"].items():
            for error in errors:
                error_type = error.get("error_type", "other")
                
                if error_type not in self.stats["errors_by_type"]:
                    self.stats["errors_by_type"][error_type] = 0
                
                self.stats["errors_by_type"][error_type] += 1

    def get_stats(self) -> Dict[str, Any]:
        """
        Get statistics about the processing.
        
        Returns:
            Dictionary of statistics
        """
        return self.stats


def main() -> int:
    """Main entry point for the incremental processor."""
    import argparse
    
    parser = argparse.ArgumentParser(
        description="Incrementally process PHPStan errors with Claude Code via MCP"
    )
    
    # Required arguments
    parser.add_argument(
        "project_path",
        help="Path to the PHP project to analyze"
    )
    
    # PHPStan options
    parser.add_argument(
        "--paths", "-p",
        nargs="+",
        help="Paths to analyze (defaults to paths in phpstan.neon)"
    )
    parser.add_argument(
        "--level", "-l",
        type=int,
        help="PHPStan rule level (0-9, defaults to level in phpstan.neon)"
    )
    parser.add_argument(
        "--config", "-c",
        help="Path to PHPStan config file (defaults to phpstan.neon in project root)"
    )
    
    # Processing options
    parser.add_argument(
        "--max-errors", "-m",
        type=int,
        default=3,
        help="Maximum errors per batch (default: 3)"
    )
    parser.add_argument(
        "--max-iterations", "-i",
        type=int,
        default=10,
        help="Maximum iterations to run (default: 10)"
    )
    parser.add_argument(
        "--verbose", "-v",
        action="store_true",
        help="Run in verbose mode"
    )
    parser.add_argument(
        "--stats-file",
        help="Path to save statistics to"
    )
    
    args = parser.parse_args()
    
    # Initialize incremental processor
    processor = IncrementalProcessor(
        args.project_path,
        args.max_errors,
        args.max_iterations,
        verbose=args.verbose
    )
    
    # Process PHPStan errors
    success = processor.process(args.paths, args.level, args.config)
    
    # Save statistics if requested
    if args.stats_file:
        stats = processor.get_stats()
        with open(args.stats_file, 'w') as f:
            json.dump(stats, f, indent=2)
    
    return 0 if success else 1


if __name__ == "__main__":
    sys.exit(main())

File: php/src/IncrementalProcessor.php
<?php

namespace MCP\PHPStan;

use MCP\PHPStan\PhpStanRunner\PhpStanRunner;
use MCP\PHPStan\ErrorFormatter\PhpStanErrorFormatter;
use MCP\PHPStan\McpIntegration\McpClient;

/**
 * Incremental PHPStan Error Processor
 *
 * This class provides functionality to incrementally process PHPStan errors
 * with Claude Code via MCP, processing a few errors at a time until all are fixed.
 */
class IncrementalProcessor
{
    private string $projectPath;
    private int $maxErrorsPerBatch;
    private int $maxIterations;
    private bool $verbose;
    private McpClient $mcpClient;
    private PhpStanRunner $phpstanRunner;
    private PhpStanErrorFormatter $errorFormatter;
    private array $stats;

    /**
     * Initialize the incremental processor
     *
     * @param string $projectPath Path to the PHP project
     * @param int $maxErrorsPerBatch Maximum number of errors to process in a single batch
     * @param int $maxIterations Maximum number of iterations to run
     * @param string|null $phpstanBinary Path to the PHPStan binary
     * @param bool $verbose Whether to run in verbose mode
     */
    public function __construct(
        string $projectPath,
        int $maxErrorsPerBatch = 3,
        int $maxIterations = 10,
        ?string $phpstanBinary = null,
        bool $verbose = false
    ) {
        $this->projectPath = realpath($projectPath);
        $this->maxErrorsPerBatch = $maxErrorsPerBatch;
        $this->maxIterations = $maxIterations;
        $this->verbose = $verbose;
        
        // Initialize MCP client
        $this->mcpClient = new McpClient($projectPath, $maxErrorsPerBatch, $phpstanBinary);
        
        // Initialize PHPStan runner
        $this->phpstanRunner = new PhpStanRunner($projectPath, $phpstanBinary);
        
        // Initialize error formatter
        $this->errorFormatter = new PhpStanErrorFormatter($maxErrorsPerBatch);
        
        // Statistics
        $this->stats = [
            'iterations' => 0,
            'total_errors_fixed' => 0,
            'total_batches_processed' => 0,
            'errors_by_type' => []
        ];
    }

    /**
     * Process PHPStan errors incrementally
     *
     * @param array|null $paths List of paths to analyze
     * @param int|null $level PHPStan rule level (0-9)
     * @param string|null $configPath Path to PHPStan config file
     * @return bool True if all errors were fixed, False otherwise
     */
    public function process(?array $paths = null, ?int $level = null, ?string $configPath = null): bool
    {
        $iteration = 0;
        $totalErrorsFixed = 0;
        
        while ($iteration < $this->maxIterations) {
            echo "\n=== Iteration " . ($iteration + 1) . "/{$this->maxIterations} ===\n";
            
            // Run PHPStan analysis
            list($output, $returnCode, $batches) = $this->mcpClient->runPhpstanAnalysis(
                $paths, 
                $level, 
                $configPath, 
                $this->verbose
            );
            
            // If no errors, we're done
            if ($returnCode === 0) {
                echo "\n✅ All PHPStan errors fixed! Total fixed: {$totalErrorsFixed}\n";
                $this->stats['iterations'] = $iteration + 1;
                $this->stats['total_errors_fixed'] = $totalErrorsFixed;
                return true;
            }
            
            // Get total errors in this iteration
            $totalErrors = array_sum(array_map(function($batch) {
                return $batch['batch']['batch_size'];
            }, $batches));
            
            echo "Found {$totalErrors} errors in " . count($batches) . " batches\n";
            
            // Process batches one by one
            $errorsFixedThisIteration = 0;
            
            foreach ($batches as $batchIdx => $batch) {
                echo "\n--- Processing batch " . ($batchIdx + 1) . "/" . count($batches) . " ---\n";
                $batchSize = $batch['batch']['batch_size'];
                
                // Prepare MCP message with file contents
                $fileContents = $this->getFileContents(array_keys($batch['errors_by_file']));
                $message = $this->mcpClient->prepareMcpMessage($batch, $fileContents);
                
                // Send to Claude
                echo "Sending batch with {$batchSize} errors to Claude...\n";
                $response = $this->mcpClient->sendToClaude($message);
                
                // Apply fixes
                if (isset($response['fixes']) && !empty($response['fixes'])) {
                    $numFixes = count($response['fixes']);
                    echo "Applying {$numFixes} fixes suggested by Claude...\n";
                    
                    $success = $this->mcpClient->applyFixes($response['fixes']);
                    if ($success) {
                        $errorsFixedThisIteration += $numFixes;
                        $this->updateErrorStats($batch);
                    } else {
                        echo "❌ Failed to apply fixes\n";
                        return false;
                    }
                } else {
                    echo "⚠️ No fixes suggested by Claude for this batch\n";
                }
                
                // Update statistics
                $this->stats['total_batches_processed']++;
                
                // Optional: Add a delay between batches to avoid rate limiting
                if ($batchIdx < count($batches) - 1) {
                    sleep(1);
                }
            }
            
            // Update total errors fixed
            $totalErrorsFixed += $errorsFixedThisIteration;
            
            // If no errors were fixed in this iteration, we're stuck
            if ($errorsFixedThisIteration === 0) {
                echo "\n⚠️ No errors fixed in iteration " . ($iteration + 1) . ". Stopping.\n";
                $this->stats['iterations'] = $iteration + 1;
                $this->stats['total_errors_fixed'] = $totalErrorsFixed;
                return false;
            }
            
            echo "\nFixed {$errorsFixedThisIteration} errors in iteration " . ($iteration + 1) . "\n";
            echo "Total errors fixed so far: {$totalErrorsFixed}\n";
            
            // Increment iteration counter
            $iteration++;
        }
        
        // If we've reached max_iterations, we've failed to fix all errors
        echo "\n⚠️ Reached maximum iterations ({$this->maxIterations}) without fixing all errors\n";
        echo "Total errors fixed: {$totalErrorsFixed}\n";
        
        $this->stats['iterations'] = $this->maxIterations;
        $this->stats['total_errors_fixed'] = $totalErrorsFixed;
        
        return false;
    }

    /**
     * Get the contents of the specified files
     *
     * @param array $filePaths List of file paths
     * @return array Dictionary mapping file paths to their contents
     */
    private function getFileContents(array $filePaths): array
    {
        $fileContents = [];
        
        foreach ($filePaths as $filePath) {
            $fullPath = $this->projectPath . '/' . $filePath;
            if (file_exists($fullPath)) {
                try {
                    $fileContents[$filePath] = file_get_contents($fullPath);
                } catch (\Exception $e) {
                    echo "Error reading file {$filePath}: {$e->getMessage()}\n";
                }
            }
        }
        
        return $fileContents;
    }

    /**
     * Update error statistics
     *
     * @param array $batch Batch of errors
     */
    private function updateErrorStats(array $batch): void
    {
        foreach ($batch['errors_by_file'] as $filePath => $errors) {
            foreach ($errors as $error) {
                $errorType = $error['error_type'] ?? 'other';
                
                if (!isset($this->stats['errors_by_type'][$errorType])) {
                    $this->stats['errors_by_type'][$errorType] = 0;
                }
                
                $this->stats['errors_by_type'][$errorType]++;
            }
        }
    }

    /**
     * Get statistics about the processing
     *
     * @return array Dictionary of statistics
     */
    public function getStats(): array
    {
        return $this->stats;
    }
}

/**
 * Main entry point for the incremental processor
 *
 * @param array $argv Command line arguments
 * @return int Exit code
 */
function main(array $argv): int
{
    // Default values
    $projectPath = $argv[1] ?? null;
    $paths = null;
    $level = null;
    $configPath = null;
    $maxErrors = 3;
    $maxIterations = 10;
    $verbose = false;
    $statsFile = null;
    
    // Check if project path is provided
    if (!$projectPath) {
        echo "Error: Project path is required\n";
        echo "Usage: php IncrementalProcessor.php <project_path> [options]\n";
        return 1;
    }
    
    // Parse command line arguments
    for ($i = 2; $i < count($argv); $i++) {
        switch ($argv[$i]) {
            case '--paths':
            case '-p':
                $paths = [];
                $i++;
                while ($i < count($argv) && !str_starts_with($argv[$i], '-')) {
                    $paths[] = $argv[$i];
                    $i++;
                }
                $i--; // Adjust for the outer loop increment
                break;
                
            case '--level':
            case '-l':
                $level = (int)$argv[++$i];
                break;
                
            case '--config':
            case '-c':
                $configPath = $argv[++$i];
                break;
                
            case '--max-errors':
            case '-m':
                $maxErrors = (int)$argv[++$i];
                break;
                
            case '--max-iterations':
            case '-i':
                $maxIterations = (int)$argv[++$i];
                break;
                
            case '--verbose':
            case '-v':
                $verbose = true;
                break;
                
            case '--stats-file':
                $statsFile = $argv[++$i];
                break;
        }
    }
    
    // Initialize incremental processor
    $processor = new IncrementalProcessor(
        $projectPath,
        $maxErrors,
        $maxIterations,
        null,
        $verbose
    );
    
    // Process PHPStan errors
    $success = $processor->process($paths, $level, $configPath);
    
    // Save statistics if requested
    if ($statsFile) {
        $stats = $processor->getStats();
        file_put_contents($statsFile, json_encode($stats, JSON_PRETTY_PRINT));
    }
    
    return $success ? 0 : 1;
}

// Only run main when script is executed directly
if (isset($argv[0]) && basename($argv[0]) === basename(__FILE__)) {
    exit(main($argv));
}
